eth/fetcher/tx_fetcher.go

```diff
@@ -943,30 +958,32 @@ func (f *TxFetcher) forEachPeer(peers map[string]struct{}, do func(peer string))
 	}
 }
 
-// forEachAnnounce does a range loop over a map of announcements in production,
-// but during testing it does a deterministic sorted random to allow reproducing
-// issues.
+// forEachAnnounce loops over the given announcements in arrival order, invoking
+// the do function for each until it returns false.
 func (f *TxFetcher) forEachAnnounce(announces map[common.Hash]*txMetadata, do func(hash common.Hash, meta *txMetadata) bool) {
-	// If we're running production, use whatever Go's map gives us
-	if f.rand == nil {
-		for hash, meta := range announces {
-			if !do(hash, meta) {
-				return
-			}
-		}
-		return
+	type announcement struct {
+		hash common.Hash
+		meta *txMetadata
 	}
-	// We're running the test suite, make iteration deterministic
-	list := make([]common.Hash, 0, len(announces))
-	for hash := range announces {
-		list = append(list, hash)
+	// process announcements by their arrival order
+	list := make([]announcement, 0, len(announces))
+	for hash, metadata := range announces {
+		list = append(list, announcement{hash: hash, meta: metadata})
 	}
-	sortHashes(list)
-	rotateHashes(list, f.rand.Intn(len(list)))
-	for _, hash := range list {
-		if !do(hash, announces[hash]) {
+	for len(list) > 0 {
+		bestArrival := math.MaxInt
+		bestIndex := -1
+		for i := range list {
+			if list[i].meta.arrival < bestArrival {
+				bestArrival = list[i].meta.arrival
+				bestIndex = i
+			}
+		}
+		if !do(list[bestIndex].hash, list[bestIndex].meta) {
```

Comment:
This is an O(n^2) loop, but since do() will exit early and the size of announcements-per-peer is capped at 2048, I'm suspecting it's typically better than doing an n*lgn sort of the entire array first.  Is it worth benchmarking?